sdk/servicebus/service-bus/test/perf/azure-sb-package/receive.ts

```diff
@@ -0,0 +1,113 @@
+/*
+# Overview
+Measures the maximum throughput of `receiver.receive()` in package `azure-sb`.
+
+# Instructions
+1. Create a Service Bus namespace with `Tier=Premium` and `Messaging Units=4`.  It is recommended to use the largest possible namespace to allow maximum client throughput.
+2. Create a queue inside the namespace.
+3. Set env vars `SERVICE_BUS_CONNECTION_STRING` and `SERVICE_BUS_QUEUE_NAME`.
+4. This test presumes that there are messages in the queue.
+5. `ts-node receive.ts [totalMessages]`
+6. Example: `ts-node receive.ts 1000000`
+ */
+
+import { createServiceBusService, ServiceBusService } from "azure-sb";
+import delay from "delay";
+import moment from "moment";
+
+const _start = moment();
+
+let _messages = 0;
+
+async function main(): Promise<void> {
+  // Endpoint=sb://<your-namespace>.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=<shared-access-key>
+  const connectionString = process.env.SERVICE_BUS_CONNECTION_STRING as string;
+  const entityPath = process.env.SERVICE_BUS_QUEUE_NAME as string;
+
+  const messages = process.argv.length > 2 ? parseInt(process.argv[2]) : 10;
+  log(`Total messages: ${messages}`);
+
+  const writeResultsPromise = WriteResults(messages);
+
+  RunTest(connectionString, entityPath, messages);
+
+  await writeResultsPromise;
+}
+
+function RunTest(connectionString: string, entityPath: string, messages: number) {
+  const sbService: ServiceBusService = createServiceBusService(connectionString);
+  function receiveMessages(done: any): void {
+    while (_messages < messages) {
```

Comment:
With a while loop here, I see this error.
![image](https://user-images.githubusercontent.com/10452642/57154302-6becad00-6d8d-11e9-878c-2cc285e3b6b4.png)
Looks like `receiveQueueMessage` is invoked too many times since the `_messages < messages` condition is satisfied.

Reply:
One way to resolve is to invoke `receiveMessages` module in the callback whenever a new message is received.
That creates a new problem of `receiveQueueMessage` being invoked too few times. 
To solve that, maybe we can add more conditional checks to limit the number of iterations of the `while` loop. 
```typescript
function RunTest(connectionString: string, entityPath: string, messages: number) {
  const sbService: ServiceBusService = createServiceBusService(connectionString);
  function receiveMessages(done: any): void {
    while (_messages < messages && count < fixed_number) {
      count++;
      sbService.receiveQueueMessage(entityPath, { isPeekLock: false }, function(err) {
        if (err) {
          console.log(err.message);
          done();
        } else {
          if (_messages === messages) {
            done();
          } else {
            _messages++;
            receiveMessages(() => {});
          }
        }
      });
    }
  }
  receiveMessages(() => {});
}
```

Reply:
Resolving this!